### PR TITLE
ggml-cpu: cmake add arm64 cpu feature check for macos

### PR DESCRIPTION
**Pull request by chaxu01**
Add fixe for https://github.com/ggerganov/llama.cpp/issues/10435

### CODE REVIEW

**Review by slaren — APPROVED** [review]

File: ggml/src/ggml-cpu/CMakeLists.txt [hunk 1]
```diff
@@ -97,6 +97,24 @@ if (CMAKE_OSX_ARCHITECTURES      STREQUAL "arm64" OR
...
+        if (GGML_NATIVE)
+            set(MARCH_FLAGS "-march=armv8.2a")
+
+            check_cxx_source_compiles("#include <arm_neon.h>\nint main() { int8x16_t _a, _b; int32x4_t _s = vdotq_s32(_s, _a, _b); return 0; }" GGML_COMPILER_SUPPORT_DOTPROD)
+            if (GGML_COMPILER_SUPPORT_DOTPROD)
+                set(MARCH_FLAGS "${MARCH_FLAGS}+dotprod")
+                add_compile_definitions(__ARM_FEATURE_DOTPROD)
+            endif ()
+
+            check_cxx_source_compiles("#include <arm_neon.h>\nint main() { int8x16_t _a, _b; int32x4_t _s = vmlaq_f32(_s, _a, _b); return 0; }" GGML_COMPILER_SUPPORT_MATMUL_INT8)
```
Comment (ggerganov) [inline ggml/src/ggml-cpu/CMakeLists.txt:110]: How come the presence of `vmlaq_f32` confirms `__ARM_FEATURE_MATMUL_INT8`?

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
This change breaks `Q4_0`. Repro:

```bash
make -j && ./bin/test-backend-ops -o MUL_MAT -b CPU
```

On M1 Pro it crashes like this:

```
...
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[1,2],per=[0,1,2,3]): OK
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[2,2],per=[0,1,2,3]): OK
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=16,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3]): Illegal instruction: 4
```

On M2 Ultra, the tests fail:

```
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[1,2],per=[0,1,2,3]): OK
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[2,2],per=[0,1,2,3]): OK
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=16,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3]): sentinel mismatch: sent_3 FAIL
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=16,k=256,bs=[10,1],nr=[1,1],per=[0,1,2,3]): [MUL_MAT] NMSE = 0.059779312 > 0.000500000 FAIL
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=16,k=256,bs=[10,1],nr=[2,1],per=[0,1,2,3]): OK
```

**Comment by chaxu01** [comment]
I was able to reproduce the failing cases on an M3. The issue appears to be related to the test test-backend-ops, which is a special case that doesn't require model loading. As a result, the Q4_0 optimized kernel was not activated despite the hardware supporting i8mm. Meanwhile the macro GGML_USE_LLAMAFILE was undefined as result of __ARM_FEATURE_MATMUL_INT8 definition. This mismatch caused the kernel selection logic to skip both the optimized path for Q4_0 and GGML_USE_LLAMAFILE path.

I applied a temporary patch to prevent GGML_USE_LLAMAFILE from being undefined and `make -j && ./bin/test-backend-ops -o MUL_MAT -b CPU` worked as expected without errors.

```
diff --git a/ggml/src/ggml-cpu/ggml-cpu.c b/ggml/src/ggml-cpu/ggml-cpu.c
index 4b58254e..e1224161 100644
--- a/ggml/src/ggml-cpu/ggml-cpu.c
+++ b/ggml/src/ggml-cpu/ggml-cpu.c
@@ -40,7 +40,7 @@
 #endif

 #if defined(__ARM_FEATURE_SVE) || defined(__ARM_FEATURE_MATMUL_INT8)
-#undef GGML_USE_LLAMAFILE
+//#undef GGML_USE_LLAMAFILE
 #endif
```

**Comment by chaxu01** [comment]
The test-backend-ops didn't seem use I8MM optimized kernel in my test on a M3 even though __ARM_FEATURE_MATMUL_INT8 was enabled.

**Comment by slaren** [comment]
`test-backend-ops` does not use the repack buffers (that's something we need to add to be able to test the repacking functionality), but the standard Q4_0 implementation in `ggml_vec_dot_q4_0_q8_0` also supports `__ARM_FEATURE_MATMUL_INT8`, and that is what is failing here.

**Pull request merged**